File: rendo-cli/cli/python/rendo_cli/project.py
from __future__ import annotations

from pathlib import Path

from .contracts import validate_project_manifest, validate_template_manifest
from .fs import read_json, write_json

WORKSPACE_METADATA_DIR = ".rendo"
PROJECT_MANIFEST_FILE = "rendo.project.json"
TEMPLATE_MANIFEST_FILE = "rendo.template.json"


def _workspace_metadata_paths(project_root: Path) -> dict[str, Path]:
    workspace_dir = project_root / WORKSPACE_METADATA_DIR
    return {
        "workspace_project": workspace_dir / PROJECT_MANIFEST_FILE,
        "workspace_template": workspace_dir / TEMPLATE_MANIFEST_FILE,
        "legacy_project": project_root / PROJECT_MANIFEST_FILE,
        "legacy_template": project_root / TEMPLATE_MANIFEST_FILE,
    }


def _mirror_workspace_metadata(project_root: Path, project: dict | None = None, template: dict | None = None) -> None:
    paths = _workspace_metadata_paths(project_root)
    if project is not None:
        write_json(paths["workspace_project"], project)
        paths["legacy_project"].unlink(missing_ok=True)

    if template is not None:
        write_json(paths["workspace_template"], template)
        paths["legacy_template"].unlink(missing_ok=True)


def _read_project_manifest(project_root: Path) -> dict | None:
    paths = _workspace_metadata_paths(project_root)
    if paths["workspace_project"].exists():
        return validate_project_manifest(read_json(paths["workspace_project"]))
    if paths["legacy_project"].exists():
        return validate_project_manifest(read_json(paths["legacy_project"]))
    return None


def _read_template_manifest(project_root: Path) -> dict | None:
    paths = _workspace_metadata_paths(project_root)
    if paths["workspace_template"].exists():
        return validate_template_manifest(read_json(paths["workspace_template"]))
    if paths["legacy_template"].exists():
        return validate_template_manifest(read_json(paths["legacy_template"]))
    return None


def find_project_root(start_dir: Path) -> Path | None:
    current = start_dir.resolve()
    while True:
        paths = _workspace_metadata_paths(current)
        if (paths["workspace_project"].exists() or paths["legacy_project"].exists()) and (
            paths["workspace_template"].exists() or paths["legacy_template"].exists()
        ):
            return current
        if current.parent == current:
            return None
        current = current.parent


def load_project_state(project_root: Path) -> tuple[dict, dict]:
    project = _read_project_manifest(project_root)
    template = _read_template_manifest(project_root)
    if project is None or template is None:
        raise RuntimeError(f"no rendo project found at {project_root}")
    return project, template


def save_project_manifest(project_root: Path, payload: dict) -> None:
    template = _read_template_manifest(project_root)
    _mirror_workspace_metadata(project_root, validate_project_manifest(payload), template)

File: rendo-cli/cli/node/src/project.ts
import { promises as fs } from "node:fs";
import path from "node:path";
import {
  projectManifestSchema,
  templateManifestSchema,
  type ProjectManifest,
  type TemplateManifest,
} from "./contracts.js";
import { pathExists, readJsonFile, writeJsonFile } from "./fs.js";

const workspaceMetadataDirName = ".rendo";
const projectManifestFileName = "rendo.project.json";
const templateManifestFileName = "rendo.template.json";

function workspaceMetadataPaths(projectRoot: string) {
  const workspaceMetadataDir = path.join(projectRoot, workspaceMetadataDirName);
  return {
    workspaceProjectManifestPath: path.join(workspaceMetadataDir, projectManifestFileName),
    workspaceTemplateManifestPath: path.join(workspaceMetadataDir, templateManifestFileName),
    legacyProjectManifestPath: path.join(projectRoot, projectManifestFileName),
    legacyTemplateManifestPath: path.join(projectRoot, templateManifestFileName),
  };
}

async function resolveWorkspaceMetadata(projectRoot: string) {
  const paths = workspaceMetadataPaths(projectRoot);
  const [
    hasWorkspaceProjectManifest,
    hasWorkspaceTemplateManifest,
    hasLegacyProjectManifest,
    hasLegacyTemplateManifest,
  ] = await Promise.all([
    pathExists(paths.workspaceProjectManifestPath),
    pathExists(paths.workspaceTemplateManifestPath),
    pathExists(paths.legacyProjectManifestPath),
    pathExists(paths.legacyTemplateManifestPath),
  ]);

  return {
    ...paths,
    hasWorkspaceProjectManifest,
    hasWorkspaceTemplateManifest,
    hasLegacyProjectManifest,
    hasLegacyTemplateManifest,
  };
}

async function readProjectManifest(metadata: Awaited<ReturnType<typeof resolveWorkspaceMetadata>>) {
  if (metadata.hasWorkspaceProjectManifest) {
    return projectManifestSchema.parse(await readJsonFile<ProjectManifest>(metadata.workspaceProjectManifestPath));
  }
  if (metadata.hasLegacyProjectManifest) {
    return projectManifestSchema.parse(await readJsonFile<ProjectManifest>(metadata.legacyProjectManifestPath));
  }
  return null;
}

async function readTemplateManifest(metadata: Awaited<ReturnType<typeof resolveWorkspaceMetadata>>) {
  if (metadata.hasWorkspaceTemplateManifest) {
    return templateManifestSchema.parse(await readJsonFile<TemplateManifest>(metadata.workspaceTemplateManifestPath));
  }
  if (metadata.hasLegacyTemplateManifest) {
    return templateManifestSchema.parse(await readJsonFile<TemplateManifest>(metadata.legacyTemplateManifestPath));
  }
  return null;
}

async function removeLegacyMetadataFiles(
  metadata: Awaited<ReturnType<typeof resolveWorkspaceMetadata>>,
  options?: { removeProject?: boolean; removeTemplate?: boolean },
) {
  if (options?.removeProject ?? true) {
    await fs.rm(metadata.legacyProjectManifestPath, { force: true });
  }
  if (options?.removeTemplate ?? true) {
    await fs.rm(metadata.legacyTemplateManifestPath, { force: true });
  }
}

export async function findProjectRoot(startDir: string): Promise<string | null> {
  let current = path.resolve(startDir);

  while (true) {
    const metadata = await resolveWorkspaceMetadata(current);
    const hasProjectManifest = metadata.hasWorkspaceProjectManifest || metadata.hasLegacyProjectManifest;
    const hasTemplateManifest = metadata.hasWorkspaceTemplateManifest || metadata.hasLegacyTemplateManifest;
    if (hasProjectManifest && hasTemplateManifest) {
      return current;
    }

    const parent = path.dirname(current);
    if (parent === current) {
      return null;
    }
    current = parent;
  }
}

export async function loadProjectState(projectRoot: string): Promise<{
  project: ProjectManifest;
  template: TemplateManifest;
}> {
  const metadata = await resolveWorkspaceMetadata(projectRoot);
  const project = await readProjectManifest(metadata);
  const template = await readTemplateManifest(metadata);
  if (!project || !template) {
    throw new Error(`no rendo project found at ${projectRoot}`);
  }
  return { project, template };
}

export async function saveProjectManifest(projectRoot: string, manifest: ProjectManifest): Promise<void> {
  const metadata = await resolveWorkspaceMetadata(projectRoot);
  const template = await readTemplateManifest(metadata);
  await writeJsonFile(metadata.workspaceProjectManifestPath, manifest);
  if (template) {
    await writeJsonFile(metadata.workspaceTemplateManifestPath, template);
  }
  await removeLegacyMetadataFiles(metadata, { removeProject: true, removeTemplate: Boolean(template) });
}

export async function readEnvKeys(envFilePath: string): Promise<Set<string>> {
  if (!(await pathExists(envFilePath))) {
    return new Set();
  }

  const content = await fs.readFile(envFilePath, "utf8");
  return new Set(
    content
      .split(/\r?\n/)
      .map((line) => line.trim())
      .filter((line) => line && !line.startsWith("#"))
      .map((line) => line.split("=", 1)[0]?.trim())
      .filter(Boolean),
  );
}
